nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java

```diff
@@ -252,4 +276,144 @@ private static boolean isIgnored(
         return false;
     }
 
+    private static final class NbJavacLoader extends URLClassLoader {
+        private static final String MAIN_COMPILER_CP = "nbjavac.class.path";
+        private static final String MAIN_COMPILER_CLASS = "com.sun.tools.javac.Main";
+        private final Map<String, Class<?>> priorityLoaded;
+
+        private NbJavacLoader(URL[] urls, ClassLoader parent) {
+            super(urls, parent);
+            this.priorityLoaded = new HashMap<>();
```

Comment:
Two questions: Does the JVM really not cache the class object? I would have expected the VM to hold the class definition in memory while the class is still in use. If it is indeed necessary, this should be a `ConcurrentHashMap` or be wrapped with `Collections#synchronizedMap`, as `loadClass` is not synchronized and there is no explicit synchronization provided there.

Reply:
I tried to cache the `Class<?>` in a static field and that isn't enough. The `CustomJavac` class is loaded again for each NetBeans project (and there are thousands of them). With:
```diff
netbeans$ git diff
diff --git nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java
index 2e17b218d4..a1a57d01b1 100644
--- nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java
+++ nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java
@@ -52,6 +52,9 @@ import org.apache.tools.ant.util.JavaEnvUtils;
  * and a separate task for {@link #cleanUpStaleClasses}.
  */
 public class CustomJavac extends Javac {
+  static {
+    System.err.println("CustomJavac loaded by " + CustomJavac.class.getClassLoader());
+  }
 
     public CustomJavac() {}
 
```
I see a lot of "CustomJavac loaded by AntClassLoader[/home/devel/NetBeansProjects/netbeans/nbbuild/build/nbantext.jar]" messages.

I'd like to share the NbJavac `Class<?>` among all such classloaders. Using (misusing) `System.getProperties` is a way to do that.

Without this trick the build was really slowed down. With the trick the time to build NetBeans with `nb-javac` is comparable to regular JDK's Javac.

Reply:
Thanks - this makes sense as I assume ant rebuilds the classloader for each javac invocation and that would explain your observation. Pulling it out into the global properties is an interesting idea to fix this.

That leaves the question regarding the missing synchronization when accessing the HashMap in a potentially racy fashion.

Reply:
> accessing the HashMap in a potentially racy fashion.

Do you mean the `HashMap` behind `Properties`? `Properties` is an ancient class and it extends `Hashtable` which is `synchronized`.

Or do you mean some other race condition that would load the `nb-javac` classes more than once? If so, then it shouldn't be a problem. The `CustomJavac` task works when loading the classes multiple times - it's just slow, so we want to cache it and not load it thousand times. As soon as the `Class` lands in global properties, any compilation that happens 1s later will pick it up. That shall be good enough.

Reply:
I mean this code:

https://github.com/apache/netbeans/blob/0112550f0a792fc6cfe0f26d74defc057a6d5cd0/nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java#L343-L354

`priorityLoaded` is a `HashMap` and if it could be accessed from multiple threads it must be synchronized.

Reply:
Thanks for your fix @matthiasblaesing in 42ed77a